File: internal/inspect/detectors/quarto.go
package detectors

// Copyright (C) 2023 by Posit Software, PBC.

import (
	"fmt"
	"path/filepath"
	"slices"
	"strings"

	"github.com/posit-dev/publisher/internal/config"
	"github.com/posit-dev/publisher/internal/contenttypes"
	"github.com/posit-dev/publisher/internal/executor"
	"github.com/posit-dev/publisher/internal/inspect/dependencies/pydeps"
	"github.com/posit-dev/publisher/internal/logging"
	"github.com/posit-dev/publisher/internal/util"
)

// These are files that behave in a special way for Quarto
// and are not included within project inspection output,
// if exists must be included in deployment files.
var specialYmlFiles = []string{
	"_quarto.yml",
	"_quarto.yaml",
	"_metadata.yml",
	"_metadata.yaml",
	"_brand.yml",
	"_brand.yaml",
}

type QuartoDetector struct {
	inferenceHelper
	executor executor.Executor
	log      logging.Logger
}

func NewQuartoDetector(log logging.Logger) *QuartoDetector {
	return &QuartoDetector{
		inferenceHelper: defaultInferenceHelper{},
		executor:        executor.NewExecutor(),
		log:             log,
	}
}

func (d *QuartoDetector) quartoInspect(path util.AbsolutePath) (*quartoInspectOutput, error) {
	args := []string{"inspect", path.String()}
	out, _, err := d.executor.RunCommand("quarto", args, util.AbsolutePath{}, d.log)
	if err != nil {
		return nil, fmt.Errorf("quarto inspect failed: %w", err)
	}
	inspectOutput, err := NewQuartoInspectOutput(out)
	if err != nil {
		return nil, err
	}
	return inspectOutput, nil
}

func (d *QuartoDetector) needsPython(inspectOutput *quartoInspectOutput) bool {
	if inspectOutput == nil {
		return false
	}
	if slices.Contains(inspectOutput.Engines, "jupyter") {
		return true
	}
	for _, script := range inspectOutput.Project.Config.Project.PreRender {
		if strings.HasSuffix(script, ".py") {
			return true
		}
	}
	for _, script := range inspectOutput.Project.Config.Project.PostRender {
		if strings.HasSuffix(script, ".py") {
			return true
		}
	}
	return false
}

func (d *QuartoDetector) needsR(inspectOutput *quartoInspectOutput) bool {
	if inspectOutput == nil {
		return false
	}
	if slices.Contains(inspectOutput.Engines, "knitr") {
		return true
	}
	for _, script := range inspectOutput.Project.Config.Project.PreRender {
		if strings.HasSuffix(script, ".R") {
			return true
		}
	}
	for _, script := range inspectOutput.Project.Config.Project.PostRender {
		if strings.HasSuffix(script, ".R") {
			return true
		}
	}
	return false
}

func (d *QuartoDetector) getTitle(inspectOutput *quartoInspectOutput, entrypointName string) string {
	isValidTitle := func(title string) bool {
		return title != "" && title != entrypointName
	}
	if isValidTitle(inspectOutput.Formats.HTML.Metadata.Title) {
		return inspectOutput.Formats.HTML.Metadata.Title
	}
	if isValidTitle(inspectOutput.Formats.RevealJS.Metadata.Title) {
		return inspectOutput.Formats.RevealJS.Metadata.Title
	}
	// Config data can exist at root level or within an additional Project object
	if isValidTitle(inspectOutput.Config.Website.Title) {
		return inspectOutput.Config.Website.Title
	}
	if isValidTitle(inspectOutput.Config.Project.Title) {
		return inspectOutput.Config.Project.Title
	}
	if isValidTitle(inspectOutput.Project.Config.Website.Title) {
		return inspectOutput.Project.Config.Website.Title
	}
	if isValidTitle(inspectOutput.Project.Config.Project.Title) {
		return inspectOutput.Project.Config.Project.Title
	}
	return ""
}

var quartoSuffixes = []string{".qmd", ".Rmd", ".ipynb", ".R", ".py", ".jl"}

func (d *QuartoDetector) findEntrypoints(base util.AbsolutePath) ([]util.AbsolutePath, error) {
	allPaths := []util.AbsolutePath{}

	for _, suffix := range quartoSuffixes {
		entrypointPaths, err := base.Glob("*" + suffix)
		if err != nil {
			return nil, err
		}
		allPaths = append(allPaths, entrypointPaths...)
	}
	return allPaths, nil
}

func isQuartoShiny(metadata *quartoMetadata) bool {
	if metadata == nil {
		return false
	}
	if metadata.Runtime == "shiny" || metadata.Server == "shiny" {
		return true
	}
	if server, ok := metadata.Server.(map[string]any); ok && server["type"] == "shiny" {
		return true
	}
	return false
}

func (d *QuartoDetector) configFromFileInspect(base util.AbsolutePath, entrypointPath util.AbsolutePath) (*config.Config, error) {
	inspectOutput, err := d.quartoInspect(entrypointPath)
	if err != nil {
		// Maybe this isn't really a quarto project, or maybe the user doesn't have quarto.
		// We log this error and continue checking the other files.
		d.log.Warn("quarto inspect failed", "file", entrypointPath.String(), "error", err)
		return nil, nil
	}

	relEntrypoint, err := entrypointPath.Rel(base)
	if err != nil {
		return nil, err
	}

	cfg := config.New()
	cfg.Entrypoint = relEntrypoint.String()
	cfg.Title = d.getTitle(inspectOutput, relEntrypoint.String())

	if isQuartoShiny(&inspectOutput.Formats.HTML.Metadata) ||
		isQuartoShiny(&inspectOutput.Formats.RevealJS.Metadata) {
		cfg.Type = contenttypes.ContentTypeQuartoShiny
	} else {
		cfg.Type = contenttypes.ContentTypeQuarto
		d.includeStaticConfig(base, cfg, inspectOutput)
	}

	var needR, needPython bool

	isDir, err := entrypointPath.IsDir()
	if err != nil {
		return nil, err
	}

	if !isDir {
		if entrypointPath.HasSuffix(".ipynb") {
			needPython = true
		} else {
			// Look for code blocks in Rmd or qmd file.
			content, err := entrypointPath.ReadFile()
			if err != nil {
				return nil, err
			}
			needR, needPython = pydeps.DetectMarkdownLanguagesInContent(content)
		}
	}

	engines := inspectOutput.Engines
	if needPython || d.needsPython(inspectOutput) {
		// Indicate that Python inspection is needed.
		cfg.Python = &config.Python{}
		if !slices.Contains(engines, "jupyter") {
			engines = append(engines, "jupyter")
		}
	}
	if needR || d.needsR(inspectOutput) {
		// Indicate that R inspection is needed.
		cfg.R = &config.R{}
		if !slices.Contains(engines, "knitr") {
			engines = append(engines, "knitr")
		}
	}
	slices.Sort(engines)

	cfg.Quarto = &config.Quarto{
		Version: inspectOutput.Quarto.Version,
		Engines: engines,
	}

	for _, inputFile := range inspectOutput.ProjectRequiredFiles() {
		var relPath string
		if filepath.IsAbs(inputFile) {
			relPath, err = filepath.Rel(base.String(), inputFile)
			if err != nil {
				return nil, err
			}
		} else {
			relPath = inputFile
		}
		cfg.Files = append(cfg.Files, fmt.Sprint("/", relPath))
	}

	for _, filename := range specialYmlFiles {
		path := base.Join(filename)
		exists, err := path.Exists()
		if err != nil {
			return nil, err
		}

		if exists {
			cfg.Files = append(cfg.Files, fmt.Sprint("/", filename))
			// Update entrypoint to keep the original _quarto.* file
			if cfg.Entrypoint == "." {
				cfg.Entrypoint = filename
			}
		}
	}
	return cfg, nil
}

// Include the static assets configuration for the Quarto project, under Config.Alternatives
// so the user has the option to deploy the static version of the project.
func (d *QuartoDetector) includeStaticConfig(base util.AbsolutePath, cfg *config.Config, inspectOutput *quartoInspectOutput) {
	entrypointRel := util.NewRelativePath(cfg.Entrypoint, nil)

	// If the entrypoint is a script, we don't generate a static configuration.
	if entrypointRel.Ext() == ".R" || entrypointRel.Ext() == ".py" {
		return
	}

	d.log.Debug("Generating Quarto static configuration", "entrypoint", cfg.Entrypoint)

	// With output-dir present, we know that any static asset will be in that directory.
	// If it is a website project, quarto inspect sets _site as the default output directory.
	if inspectOutput.OutputDir() != "" {
		staticCfg := d.staticConfigFromOutputDir(base, cfg, inspectOutput)
		if staticCfg != nil {
			cfg.Alternatives = append(cfg.Alternatives, *staticCfg)
		}
		return
	}

	d.log.Debug("Quarto project does not specify an output-dir")
	d.log.Debug("Attempting to generate Quarto document static configuration from HTML files in project")

	// Last option to generate a static configuration,
	// use inspect output files.input to generate a list with the .html ext version of the files.
	// (If there is a render list, files.input is already filtered down to the applicable files.)
	staticCfg := d.staticConfigFromFilesLookup(base, cfg, inspectOutput)
	if staticCfg != nil {
		cfg.Alternatives = append(cfg.Alternatives, *staticCfg)
	}
}

// Generate a static configuration based on the output-dir specified in the project,
// with discoverability and generation of the entrypoint HTML file related to the output-dir.
func (d *QuartoDetector) staticConfigFromOutputDir(base util.AbsolutePath, cfg *config.Config, inspectOutput *quartoInspectOutput) *config.Config {
	outputDir := inspectOutput.OutputDir()

	d.log.Debug("Quarto project has output-dir specified", "output_dir", outputDir)

	staticCfg := config.New()
	staticCfg.Type = contenttypes.ContentTypeHTML
	staticCfg.Title = cfg.Title

	// Prep the rendered version of the entrypoint path.
	outputDirRel := util.NewRelativePath(outputDir, nil)
	outputDirEntrypoint := outputDirRel.Join(cfg.Entrypoint)
	htmlVerEntrypoint := outputDirEntrypoint.WithoutExt().String() + ".html"

	entrypointIsDir, err := base.Join(cfg.Entrypoint).IsDir()
	if err != nil {
		d.log.Error("Error generating Quarto static configuration", "error", err)
		return nil
	}

	// If the entrypoint is a directory or _quarto.yml:
	// - look up for an index file
	// - OR use the static version of the first file in files.input, "outputDir + files[0] + .html"
	if d.isQuartoYaml(cfg.Entrypoint) || entrypointIsDir {
		d.log.Debug("Choosen entrypoint is _quarto.yml, attemtping to use first file from files.input as static entrypoint")

		indexFile, foundIndex := inspectOutput.IndexHTMLFilepath(base)

		if !foundIndex {
			htmlInputFiles := inspectOutput.HTMLAbsPathsFromInputList(base)
			indexFile = htmlInputFiles[0]
		}

		indexFileRel, err := indexFile.Rel(base)
		if err != nil {
			d.log.Error("Error generating Quarto static configuration", "error", err)
			return nil
		}

		// htmlVerEntrypoint = htmlInputFiles[0].String()
		htmlVerEntrypoint = outputDirRel.Join(indexFileRel.String()).String()
		d.log.Debug("Using first file from files.input as static entrypoint", "entrypoint", htmlVerEntrypoint)
	}

	staticCfg.Entrypoint = htmlVerEntrypoint
	staticCfg.Files = []string{fmt.Sprint("/", outputDir)}
	return staticCfg
}

// Generate a static configuration based on the existing project files that quarto inspect reports.
func (d *QuartoDetector) staticConfigFromFilesLookup(base util.AbsolutePath, cfg *config.Config, inspectOutput *quartoInspectOutput) *config.Config {
	staticCfg := config.New()
	staticCfg.Type = contenttypes.ContentTypeHTML
	staticCfg.Title = cfg.Title

	htmlInputFiles := inspectOutput.HTMLAbsPathsFromInputList(base)
	for _, file := range htmlInputFiles {
		relFile, err := file.Rel(base)
		if err != nil {
			d.log.Error("Error generating Quarto static configuration", "error", err)
			return nil
		}

		// If the entrypoint is not set, use the first file provided by quarto as the entrypoint.
		if staticCfg.Entrypoint == "" {
			staticCfg.Entrypoint = relFile.Base()
		}
		staticCfg.Files = append(staticCfg.Files, fmt.Sprint("/", relFile.String()))

		// Include any accompanying *_files directory for each HTML file.
		assetsDir, existsDir := inspectOutput.fileAssetsDir(file)
		if existsDir {
			relAssetsDir, err := assetsDir.Rel(base)
			if err != nil {
				d.log.Error("Error generating Quarto static configuration", "error", err)
				return nil
			}
			staticCfg.Files = append(staticCfg.Files, fmt.Sprint("/", relAssetsDir.String()))
		}
	}

	// If it wasn't possible to find any files for a static configuration,
	// we don't include alternatives.
	if len(staticCfg.Files) > 0 {
		return staticCfg
	}
	return nil
}

func (d *QuartoDetector) isQuartoYaml(entrypointBase string) bool {
	return slices.Contains([]string{"_quarto.yml", "_quarto.yaml"}, entrypointBase)
}

func (d *QuartoDetector) InferType(base util.AbsolutePath, entrypoint util.RelativePath) ([]*config.Config, error) {
	// When the choosen entrypoint is _quarto.yml, "quarto inspect" command does not handle it well
	// but an inspection to the base directory will bring what the user expects in this case.
	if d.isQuartoYaml(entrypoint.Base()) {
		d.log.Debug("A _quarto.yml file was picked as entrypoint", "inspect_path", base.String())

		cfg, err := d.configFromFileInspect(base, base)
		if err != nil {
			return nil, err
		}

		if cfg != nil {
			return []*config.Config{cfg}, nil
		}
	}

	if entrypoint.String() != "" {
		// Optimization: skip inspection if there's a specified entrypoint
		// and it's not one of ours.
		if !slices.Contains(quartoSuffixes, entrypoint.Ext()) {
			d.log.Debug("Picked entrypoint does not match Quarto file extensions, skipping", "entrypoint", entrypoint.String())
			return nil, nil
		}
	}

	var configs []*config.Config
	entrypointPaths, err := d.findEntrypoints(base)
	if err != nil {
		return nil, err
	}

	for _, entrypointPath := range entrypointPaths {
		relEntrypoint, err := entrypointPath.Rel(base)
		if err != nil {
			return nil, err
		}
		if entrypoint.String() != "" && relEntrypoint != entrypoint {
			// Only inspect the specified file
			continue
		}

		cfg, err := d.configFromFileInspect(base, entrypointPath)
		if err != nil {
			return nil, err
		}

		if cfg != nil {
			configs = append(configs, cfg)
		}
	}
	return configs, nil
}

File: extensions/vscode/src/inspect/detectors/quarto.ts
// Copyright (C) 2026 by Posit Software, PBC.

import * as path from "path";
import * as fs from "fs/promises";
import { execFile } from "child_process";
import { promisify } from "util";
import { ContentType } from "src/api/types/configurations";
import { logger } from "src/logging";
import { ContentTypeDetector, PartialConfig } from "../types";
import { globDir } from "../helpers/globDir";
import { detectMarkdownLanguagesInContent } from "../helpers/markdownLanguages";
import { findLinkedResources } from "../helpers/resourceFinder";
import { QuartoInspectOutput } from "./quartoInspectOutput";

const execFileAsync = promisify(execFile);

const specialYmlFiles = [
  "_quarto.yml",
  "_quarto.yaml",
  "_metadata.yml",
  "_metadata.yaml",
  "_brand.yml",
  "_brand.yaml",
];

// Fallback version used when `quarto inspect` is unavailable (e.g. quarto not
// installed). Matches the latest stable release at the time this detector was
// written. The exact value is non-critical — Connect uses it only as a hint.
const defaultQuartoVersion = "1.7.34";

const quartoSuffixes = [".qmd", ".Rmd", ".ipynb", ".R", ".py", ".jl"];
const quartoSuffixesLower = quartoSuffixes.map((s) => s.toLowerCase());

function isQuartoShiny(metadata: {
  runtime?: string;
  server?: unknown;
}): boolean {
  if (metadata.runtime === "shiny" || metadata.server === "shiny") {
    return true;
  }
  if (
    typeof metadata.server === "object" &&
    metadata.server !== null &&
    "type" in metadata.server &&
    metadata.server.type === "shiny"
  ) {
    return true;
  }
  return false;
}

function isQuartoYaml(entrypointBase: string): boolean {
  return ["_quarto.yml", "_quarto.yaml"].includes(entrypointBase);
}

function isExpectedInspectFailure(err: unknown): boolean {
  if (typeof err === "object" && err !== null) {
    const obj = err as Record<string, unknown>;
    if (obj.code === "ENOENT") {
      return true;
    }
    if (typeof obj.stderr === "string") {
      return obj.stderr.includes("is not a valid Quarto input document");
    }
  }
  return false;
}

export class QuartoDetector implements ContentTypeDetector {
  async inferType(
    baseDir: string,
    entrypoint?: string,
  ): Promise<PartialConfig[]> {
    // The `quarto inspect` CLI does not accept _quarto.yml as a direct target.
    // Instead, pass the project directory which inspects the full project.
    if (entrypoint && isQuartoYaml(path.basename(entrypoint))) {
      const cfg = await this.configFromInspect(baseDir, baseDir);
      if (cfg) {
        // Override entrypoint to the directory dot (will be resolved later)
        cfg.entrypoint = ".";
        return [cfg];
      }
    }

    if (entrypoint) {
      const ext = path.extname(entrypoint).toLowerCase();
      if (!quartoSuffixesLower.includes(ext)) {
        return [];
      }
    }

    const configs: PartialConfig[] = [];
    const entrypointPaths = await this.findEntrypoints(baseDir);

    for (const epPath of entrypointPaths) {
      const relEntrypoint = path.basename(epPath);
      if (entrypoint && relEntrypoint !== entrypoint) {
        continue;
      }
      const cfg = await this.configFromInspect(baseDir, epPath);
      if (cfg) {
        configs.push(cfg);
      }
    }
    return configs;
  }

  private async quartoInspect(
    inspectPath: string,
  ): Promise<QuartoInspectOutput> {
    const { stdout } = await execFileAsync("quarto", ["inspect", inspectPath], {
      timeout: 30_000,
    });
    return new QuartoInspectOutput(stdout);
  }

  private async configFromInspect(
    baseDir: string,
    inspectPath: string,
  ): Promise<PartialConfig | undefined> {
    let inspectOutput: QuartoInspectOutput;
    try {
      inspectOutput = await this.quartoInspect(inspectPath);
      logger.debug(`[quarto] quarto inspect succeeded for ${inspectPath}`);
    } catch (err: unknown) {
      const expected = isExpectedInspectFailure(err);
      const log = expected ? logger.debug : logger.warn;
      log(
        `[quarto] quarto inspect failed for ${inspectPath}, attempting fallback: ${err}`,
      );
      const fallbackCfg = await this.genNonInspectConfig(baseDir, inspectPath);
      if (fallbackCfg) {
        logger.info(
          `[quarto] generated configuration without quarto binary inspection: ${inspectPath}`,
        );
      } else {
        log(
          `[quarto] could not identify Quarto project by files context: ${inspectPath}`,
        );
      }
      return fallbackCfg;
    }

    const relEntrypoint =
      inspectPath === baseDir ? "." : path.basename(inspectPath);

    const cfg: PartialConfig = {
      type: ContentType.QUARTO_STATIC,
      entrypoint: relEntrypoint,
    };

    cfg.title = inspectOutput.getTitle(relEntrypoint);

    // Determine content type: shiny or static
    if (
      isQuartoShiny(inspectOutput.htmlMetadata) ||
      isQuartoShiny(inspectOutput.revealjsMetadata)
    ) {
      logger.info(`[quarto] detected Quarto Shiny content: ${relEntrypoint}`);
      cfg.type = ContentType.QUARTO_SHINY;
    }

    // Detect language needs
    let needR = false;
    let needPython = false;

    const isDir = inspectPath === baseDir;
    if (!isDir) {
      if (inspectPath.endsWith(".ipynb")) {
        needPython = true;
      } else {
        try {
          const content = await fs.readFile(inspectPath, "utf-8");
          const langs = detectMarkdownLanguagesInContent(content);
          needR = langs.needsR;
          needPython = langs.needsPython;
        } catch (err: unknown) {
          logger.debug(
            `[quarto] could not read file for language detection: ${inspectPath}: ${err}`,
          );
        }
      }
    }

    const engines = [...inspectOutput.engines];
    if (needPython || this.needsPython(inspectOutput)) {
      cfg.python = {};
      if (!engines.includes("jupyter")) {
        engines.push("jupyter");
      }
    }
    if (needR || this.needsR(inspectOutput)) {
      cfg.r = {};
      if (!engines.includes("knitr")) {
        engines.push("knitr");
      }
    }
    engines.sort();

    cfg.quarto = {
      version: inspectOutput.version,
      engines,
    };

    // Include project files
    cfg.files = await this.collectProjectFiles(baseDir, cfg, inspectOutput);

    // Discover linked resources (images, stylesheets, etc.)
    const discoveredAssets = await findLinkedResources(baseDir, cfg.files);
    cfg.files.push(...discoveredAssets);

    // Include static alternative for non-Shiny content
    if (cfg.type !== ContentType.QUARTO_SHINY) {
      const alt = await this.buildStaticAlternative(
        baseDir,
        cfg,
        inspectOutput,
      );
      if (alt) {
        logger.debug(
          `[quarto] generated static alternative with entrypoint: ${alt.entrypoint}`,
        );
        cfg.alternatives = [alt];
      }
    }

    return cfg;
  }

  private needsPython(inspectOutput: QuartoInspectOutput): boolean {
    if (inspectOutput.engines.includes("jupyter")) {
      return true;
    }
    for (const script of inspectOutput.prePostRenderFiles()) {
      if (script.toLowerCase().endsWith(".py")) {
        return true;
      }
    }
    return false;
  }

  private needsR(inspectOutput: QuartoInspectOutput): boolean {
    if (inspectOutput.engines.includes("knitr")) {
      return true;
    }
    for (const script of inspectOutput.prePostRenderFiles()) {
      if (script.toLowerCase().endsWith(".r")) {
        return true;
      }
    }
    return false;
  }

  private async collectProjectFiles(
    baseDir: string,
    cfg: PartialConfig,
    inspectOutput: QuartoInspectOutput,
  ): Promise<string[]> {
    const files: string[] = [];

    for (const inputFile of inspectOutput.projectRequiredFiles()) {
      const relPath = path.isAbsolute(inputFile)
        ? path.relative(baseDir, inputFile)
        : inputFile;

      // Skip files within _extensions directory (handled separately)
      if (relPath.startsWith("_extensions")) {
        continue;
      }
      files.push(`/${relPath}`);
    }

    await this.includeSpecialYmlFiles(baseDir, files, cfg);
    await this.includeExtensionsDir(baseDir, files);

    return files;
  }

  private async includeSpecialYmlFiles(
    baseDir: string,
    files: string[],
    cfg: PartialConfig,
  ): Promise<void> {
    for (const filename of specialYmlFiles) {
      const filePath = path.join(baseDir, filename);
      try {
        await fs.access(filePath);
        files.push(`/${filename}`);
        // Update entrypoint to keep the original _quarto.* file
        if (cfg.entrypoint === ".") {
          cfg.entrypoint = filename;
        }
      } catch {
        // File doesn't exist — not an error, just skip
      }
    }
  }

  private async includeExtensionsDir(
    baseDir: string,
    files: string[],
  ): Promise<void> {
    const extensionsDir = path.join(baseDir, "_extensions");
    try {
      const stat = await fs.stat(extensionsDir);
      if (stat.isDirectory()) {
        logger.debug("[quarto] including _extensions directory");
        files.push("/_extensions");
      }
    } catch (err: unknown) {
      logger.debug(`[quarto] could not check _extensions directory: ${err}`);
    }
  }

  // Builds an HTML alternative config that represents the pre-rendered static
  // output of a Quarto project. Two strategies are tried:
  //   1. If the project declares an output-dir (e.g. "_site"), point at that
  //      directory and resolve the HTML entrypoint within it.
  //   2. Otherwise, derive HTML filenames from the input file list (e.g.
  //      "doc.qmd" → "doc.html").
  private async buildStaticAlternative(
    baseDir: string,
    cfg: PartialConfig,
    inspectOutput: QuartoInspectOutput,
  ): Promise<PartialConfig | undefined> {
    const ext = path.extname(cfg.entrypoint).toLowerCase();
    // Script entrypoints (.R, .py) don't produce standalone HTML output
    if (ext === ".r" || ext === ".py") {
      return undefined;
    }

    const outputDir = inspectOutput.outputDir();
    if (outputDir) {
      return this.staticConfigFromOutputDir(baseDir, cfg, inspectOutput);
    }

    return await this.staticConfigFromFilesLookup(baseDir, cfg, inspectOutput);
  }

  // Strategy 1: The project has an explicit output-dir (e.g. "_site" for
  // website projects). The whole directory is included as a file entry, and
  // the HTML entrypoint is resolved differently depending on whether the
  // source entrypoint is a directory/yaml config or a regular file.
  private staticConfigFromOutputDir(
    baseDir: string,
    cfg: PartialConfig,
    inspectOutput: QuartoInspectOutput,
  ): PartialConfig | undefined {
    const outputDir = inspectOutput.outputDir();

    const staticCfg: PartialConfig = {
      type: ContentType.HTML,
      entrypoint: "",
      title: cfg.title,
      source: this.renderSource(cfg),
      files: [`/${outputDir}`],
    };

    // For directory/yaml entrypoints, find an index or first input file
    if (isQuartoYaml(cfg.entrypoint) || cfg.entrypoint === ".") {
      const indexFile = inspectOutput.indexHTMLFilepath(baseDir);
      if (indexFile) {
        const relIndex = path.relative(baseDir, indexFile);
        staticCfg.entrypoint = path.join(outputDir, relIndex);
      } else {
        const htmlFiles = inspectOutput.htmlPathsFromInputList(baseDir);
        if (htmlFiles.length > 0) {
          const firstHtml = htmlFiles[0];
          if (firstHtml) {
            const relFirst = path.relative(baseDir, firstHtml);
            staticCfg.entrypoint = path.join(outputDir, relFirst);
          }
        }
      }
    } else {
      // Standard entrypoint: compute its HTML version in output dir
      const stem = path.basename(cfg.entrypoint, path.extname(cfg.entrypoint));
      staticCfg.entrypoint = path.join(outputDir, stem + ".html");
    }

    return staticCfg;
  }

  // Strategy 2: No output-dir declared. Derive HTML filenames directly from
  // the input file list by replacing each file's extension with ".html".
  // The first input file becomes the entrypoint.
  private async staticConfigFromFilesLookup(
    baseDir: string,
    cfg: PartialConfig,
    inspectOutput: QuartoInspectOutput,
  ): Promise<PartialConfig | undefined> {
    const files: string[] = [];
    let entrypoint = "";

    const inputFiles = inspectOutput.inputFiles();
    for (const file of inputFiles) {
      const stem = path.basename(file, path.extname(file));
      const htmlFile = stem + ".html";

      if (!entrypoint) {
        entrypoint = htmlFile;
      }
      files.push(`/${htmlFile}`);

      // Include any accompanying *_files directory for each HTML file.
      const htmlAbsPath = path.join(baseDir, htmlFile);
      const assetsDir = await inspectOutput.fileAssetsDir(htmlAbsPath);
      if (assetsDir) {
        const relAssetsDir = path.relative(baseDir, assetsDir);
        logger.debug(`[quarto] including companion directory: ${relAssetsDir}`);
        files.push(`/${relAssetsDir}`);
      }
    }

    if (files.length === 0) {
      return undefined;
    }

    const discoveredAssets = await findLinkedResources(baseDir, files);
    files.push(...discoveredAssets);

    return {
      type: ContentType.HTML,
      entrypoint,
      title: cfg.title,
      source: this.renderSource(cfg),
      files,
    };
  }

  private renderSource(cfg: PartialConfig): string {
    if (cfg.entrypoint === ".") {
      return "_quarto.yml";
    }
    return cfg.entrypoint;
  }

  private async findEntrypoints(baseDir: string): Promise<string[]> {
    const results = await Promise.all(
      quartoSuffixes.map((suffix) =>
        globDir(baseDir, "*" + suffix, { nocase: true }),
      ),
    );

    return results.flat();
  }

  private async genNonInspectConfig(
    baseDir: string,
    inspectPath: string,
  ): Promise<PartialConfig | undefined> {
    // Check if _quarto.yml or _quarto.yaml exists
    let quartoYmlExists = false;
    for (const name of ["_quarto.yml", "_quarto.yaml"]) {
      try {
        await fs.access(path.join(baseDir, name));
        quartoYmlExists = true;
        break;
      } catch {
        // doesn't exist
      }
    }

    const ext = path.extname(inspectPath).toLowerCase();
    if (
      !quartoYmlExists &&
      ext !== ".qmd" &&
      ext !== ".ipynb" &&
      ext !== ".rmd"
    ) {
      return undefined;
    }

    const relEntrypoint =
      inspectPath === baseDir ? "." : path.basename(inspectPath);

    logger.debug(
      `[quarto] attempting fallback file-based detection for ${relEntrypoint}`,
    );

    const files: string[] = [];
    const cfg: PartialConfig = {
      type: ContentType.QUARTO_STATIC,
      entrypoint: relEntrypoint,
      quarto: {
        version: defaultQuartoVersion,
      },
      files,
    };

    // Standalone Jupyter notebooks need Python and the jupyter engine
    if (ext === ".ipynb") {
      cfg.python = {};
      cfg.quarto = { version: defaultQuartoVersion, engines: ["jupyter"] };
      files.push(`/${relEntrypoint}`);
      const assets = await findLinkedResources(baseDir, files);
      files.push(...assets);
      return cfg;
    }

    // Standalone RMarkdown files need R and the knitr engine
    if (ext === ".rmd") {
      cfg.r = {};
      cfg.quarto = { version: defaultQuartoVersion, engines: ["knitr"] };
      files.push(`/${relEntrypoint}`);
      const assets = await findLinkedResources(baseDir, files);
      files.push(...assets);
      return cfg;
    }

    // Include .qmd files
    const qmdFiles = await globDir(baseDir, "*.qmd");
    for (const qmdPath of qmdFiles) {
      const relPath = path.basename(qmdPath);
      files.push(`/${relPath}`);
    }

    // Include special yml files
    await this.includeSpecialYmlFiles(baseDir, files, cfg);

    const assets = await findLinkedResources(baseDir, files);
    files.push(...assets);

    return cfg;
  }
}
